Rx build fixes for SDK 0.18

keith-packard opened the pull request:
I'm just trying to test SDK version 0.17.1 with RX support; these fixes are helping get things building, but I haven't a clue if most of them are correct.

stephanosio committed:
cmake: Rename 'zephyr' toolchain variant to 'zephyr-gnu'

This commit renames the existing `zephyr` toolchain variant, which is used
to select the Zephyr SDK GNU (Binutils and GCC) toolchain to `zephyr-gnu`
in preparation for introducing the Zephyr SDK LLVM toolchain variant.

After this change, the `ZEPHYR_TOOLCHAIN_VARIANT` environment variable must
be set to `zephyr-gnu` in order to select the Zephyr SDK GNU toolchain.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
cmake: Deprecate 'zephyr' toolchain variant gracefully

After the renaming of the `zephyr` toolchain variant to `zephyr-gnu`, this
commit gracefully deprecates the `zephyr` toolchain variant by displaying a
warning message and handling the legacy Zephyr SDK directory layout when
`ZEPHYR_TOOLCHAIN_VARIANT=zephyr`.

This effectively allows legacy Zephyr SDK releases (<= 0.17) to be used
as-is in the current version of Zephyr.

For Zephyr SDK >= 0.18, which has the new directory layout, the
`zephyr-gnu` toolchain variant must be selected.

Revert this commit when the legacy `zephyr` toolchain variant support is
removed.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
ci: Update deprecated 'zephyr' toolchain variant references

This commit updates all references to the `zephyr` toolchain variant, which
has been deprecated, to `zephyr-gnu` in the CI workflows.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
doc: Update deprecated 'zephyr' toolchain variant references

This commit updates all references to the `zephyr` toolchain variant, which
has been deprecated, to `zephyr-gnu` in the documentation.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kconfig: Update deprecated 'zephyr' toolchain variant references

This commit updates all references to the `zephyr` toolchain variant, which
has been deprecated, to `zephyr-gnu` in the Kconfig files and build
scripts.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kconfig: Deprecate 'zephyr' toolchain variant gracefully

After the renaming of the `zephyr` toolchain variant to `zephyr-gnu`, this
commit gracefully deprecates the `zephyr` toolchain variant by treating
`zephyr` toolchain variant to be equivalent to `zephyr-gnu` toolchain
variant.

Revert this commit when the legacy `zephyr` toolchain variant support is
removed.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
cmake: Add 'zephyr-llvm' toolchain variant

This commit adds a new toolchain variant `zephyr-llvm` to be used with the
Zephyr SDK LLVM toolchain.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
lib: runtime: Select COMPILER_RT_SUPPORTED for zephyr-llvm toolchain

Select `COMPILER_RT_SUPPORTED` for `zephyr-llvm` toolchain variant since
Zephyr SDK LLVM toolchain includes pre-built compiler-rt library.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
Kconfig.zephyr: Default to aggressive size optimisation for zephyr-llvm

Default to aggressive size optimisation (`SIZE_OPTIMIZATIONS_AGGRESSIVE`)
for Zephyr SDK LLVM toolchain as with the generic LLVM toolchain.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
cmake: toolchain: Introduce TOOLCHAIN_HAS_GLIBCXX

This commit introduces `TOOLCHAIN_HAS_GLIBCXX` CMake variable, which is
set to `y` when GNU C++ Standard Library aka. libstdc++ is available.

This helps filter libstdc++-specific Kconfig and tests in a more refined
manner.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: libcxx: Filter libstdc++ tests by TOOLCHAIN_HAS_GLIBCXX

Run libstdc++ tests only using the toolchains that have libstdc++ (i.e.
GCC-based toolchain).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
cmake: toolchain: Introduce TOOLCHAIN_HAS_LIBCXX

This commit introduces `TOOLCHAIN_HAS_LIBCXX` CMake variable, which is set
to `y` when LLVM C++ Standard Library aka. libc++ is available.

This helps filter libc++-specific Kconfig and tests in a more refined
manner.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
lib: cpp: Allow selecting libc++ with picolibc

LLVM C++ Standard Library aka. libc++ may be used with Picolibc -- for
example, LLVM Embedded Toolchain for Arm ships with Picolibc and libc++
compiled for Picolibc, and so does Zephyr SDK LLVM toolchain.

Ideally, we would a flag like `TOOLCHAIN_HAS_LIBCXX_PICOLIBC` indicating
that toolchain has libc++ compiled specifically for Picolibc; but, we do
not want to overcomplicate the toolchain feature flags at this time without
proper underlying infrastructure to handle it.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: libcxx: Add LLVM libc++/picolibc test

This commit adds a new C++ standard library test variant that builds with
Picolibc and LLVM C++ Standard Library aka. libc++ on the supported
toolchains (e.g. LLVM Embedded Toolchain for Arm and Zephyr SDK LLVM).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
lib: cpp: Enable POSIX and GNU C extensions for libc++

LLVM C++ Standard Library aka. libc++ makes use of POSIX and GNU C
extensions in its headers.

While far from ideal, there is no other option but to globally enable these
extensions for C++ source files in order to ensure that libc++ headers can
be used.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
cmake: linker: lld: Do not link libc when minimal libc is used

When LLVM linker is used, the toolchain-provided C standard library is
always linked, even when the Zephyr minimal libc is selected, because `c`
is unconditionally specified in `link_order_library`, which causes `-lc` to
be specified in the linker flags.

This commit adds a check to ensure that `-lc` is not specified when the
Zephyr minimal libc is selected because it is a standalone C standard
library and it does not make sense to link two C standard libraries at
once.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

keith-packard committed:
cmake: Pass OPTIMIZATION_FLAG to linker too

The compiler may want to know the desired optimization level during
linking, as when the compiler multilib configuration includes -Os as a
selector.

Do this by adding a new (optional) linker function,
toolchain_linker_add_compiler_options, which maps compiler options to
equivalent linker options, discarding any that aren't applicable.

Linker configurations which use the compiler driver (ld, lld, and xt-ld)
apply all provided compiler options. Linkers which don't provide this
function fall back to an implementation which simply discards all options.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
cmake: Delay computation of linker paths until needed

With inclusion of the optimization flag into the multilib selection
process, we cannot compute the compiler library path when the compiler's
target.cmake is processed as OPTIMIZATION_FLAG is not computed until much
later.

Instead, add a function (compiler_file_path) which can be used to locate
the appropriate crtbegin.o and crtend.o files.

Delay computation of lib_include_dir and rt_library until after all
compiler flags have been computed by adding compiler_set_linker_properties
and calling that just before toolchain_linker_finalize is invoked.

Place default implementations of both of these functions in a new file,
cmake/compiler/target_template.cmake, where we assume the compiler works
like gcc or clang and handlers the --print-file-name and
--print-libgcc-file-name options. Compilers needing alternate
implementations can override these functions in their target.cmake files.

These implementations require that no generator expressions are necessary
for the compiler to compute the right library paths.

This mechanism is also used to take any additional compiler options and
pass them to the linker using toolchain_linker_add_compiler_options.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
arch/rx: Add missing linker script fragments

There were lots of missing chunks, including those necessary for C++
support.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
boards: Add zephyr toolchain to rx boards

SDK version 0.17.1 and later include RX support, so include the zephyr
toolchain in the board files using these processors.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/arch/rx: Remove unused variable 'events'

This causes a compile error.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
subsys/debug: Stub out debug thread info for RX arch

This architecture is missing lots of support bits. Stub out this piece so
we can get more things building.

Signed-off-by: Keith Packard <keithp@keithp.com>

nashif committed:
tests: c_lib: exclude rx arch

Many tests failing for RX, exclude those until the arch stablizes.

See #89839

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

keith-packard committed:
posix: Use 'char' for posix_condattr bitfields

Ensure that posix_condattr is smaller than pthread_condattr by keeping it
under one byte.

Signed-off-by: Keith Packard <keithp@keithp.com>

nashif commented on boards/qemu/rx/qemu_rx.yaml:
```diff
@@ -9,6 +9,7 @@ simulation:
 arch: rx
 toolchain:
   - cross-compile
+  - zephyr
```
too early for that, need to wait for 0.17.1 to be released

keith-packard committed:
scripts/logging: Allow targets that prefix global symbols with '_'

RX adds an underscore to all global symbols, so when we look for
log_const_ globals, we need to allow for that.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
scripts/logging: Skip log symbols of zero size

I'm not sure why this was necessary, but analyzing RX
executables was crashing with a divide by zero error on a
symbol which appeared to be duplicated in this list.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
scripts/logging: Add RX config detection

The RX arch uses the CONFIG_RX kconfig symbol

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
subsys/mgmt: Declare PROCESSOR_NAME for RX targets

Map the current set of RX CPU variants into descriptive names

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/minimal: Add stub 'sys/lock.h'

This file is explicitly included by the espressif hal module. It's an
internal file provided by picolibc and newlib. Provide a stub to let
code designed for those to work with the minimal C library.

Signed-off-by: Keith Packard <keithp@keithp.com>

nashif committed:
boards: Update 'zephyr' toolchain variant references to 'zephyr-gnu'

This commit updates all deprecated `zephyr` toolchain variant references in
the board YAML files to `zephyr-gnu`.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

keith-packard committed:
modules/cmsis-dsp: Don't use Zephyr stdint.h

Zephyr replaces the toolchain version of stdint.h to define uint32_t and
int32_t as int rather than long. This breaks the ARM MVE intrinics which
require uint32_t to be defined as unsigned long.

Define 'ZEPHYR_USE_TOOLCHAIN_STDINT_H_' as that causes zephyr_stdint.h to
be skipped, ensuring that the normal stdint.h types are used instead.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
arch/x86: Support picolibc with zefi

Disable the default startup file with -nostartfiles.

Signed-off-by: Keith Packard <keithp@keithp.com>

nashif committed:
cmake: clang: -Wvolatile is deprecated

-Wvolatile is deprecated

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

keith-packard committed:
Revert "tests: c_lib: exclude rx arch"

This reverts commit 7762b560ba6b5b3471bed20e84a753ed4edf5ed9.

keith-packard committed:
boards/qemu/rx: Add zephyr-gnu toolchain

Adapt to the toolchain name change

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
boards/renesas: adapt to zephyr SDK toolchain name change

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
compiler/gcc: Add rx compiler flags to linker too

Many of these affect the library search path.

Signed-off-by: Keith Packard <keithp@keithp.com>
